Add docs.rs opt-in single target post (#548)
rust-lang/blog.rust-lang.org: posts/2020-03-15-docs-rs-opt-into-fewer-targets.md

---

Recently, [docs.rs] added a feature that allows crates to opt-out of building on all targets.
If you don't need to build on all targets, we ask that you enable this feature
to allow us to reduce queue times and storage costs.

## What does the feature do?

By default, docs.rs builds all crates published to [crates.io] for [every tier one target][metadata].
However, most crates have the same content on all target.
Of the platform-dependent crates, almost all target a single platform,
and do not need to be built on other targets.
For example, [`winapi`] only has documentation on the `x86_64-pc-windows-msvc`
and `i686-pc-windows-msvc` targets, and is blank on all others.

This feature allows you to request building only on specific targets.
For example, [`winapi`] could opt into only building windows targets
by putting the following in its `Cargo.toml`:

```toml
[package.metadata.docs.rs]
# This also sets the default target to `x86_64-pc-windows-msvc`
targets = ["x86_64-pc-windows-msvc", "i686-pc-windows-msvc"]
```

If you only need a single target, it's even simpler:

```toml
[package.metadata.docs.rs]
# This builds only the default target (usually `x86_64-unknown-linux-gnu`)

> targets = []

Reviewer: I would not recommend using the empty target list. While it's supported to prevent unnecessary build failures, we should encourage people to fill the default target in if they use `targets`.

Author: I think this will help increase adoption by making it easier to opt-in, though. If we change it to suggest `targets = ["x86_64-unknown-linux-gnu"]`, then someone might type it by hand and spell it wrong, failing the build for no reason.

Reviewer: I don't think this will prevent adoption too much, as people will copy/paste this snippet anyway. Also, your misspelling concern would apply to every other target people could add.

Author: Ok, I'll change it to `targets = ["x86_64-unknown-linux-gnu"]`.

```

See the [docs.rs documentation][metadata] for more details about how to opt-in.

## How does this help docs.rs?

Building all crates from crates.io can take a long time!
Building fewer targets will allow us to reduce wait times for every crate.

> This will avoid delaying every crate by hours when e.g. rusoto publishes a new release.

Reviewer: Let's not "blame" crates if possible.

Additionally, this will decrease our fixed storage costs,
allowing docs.rs to be sustainable into the future.

## Why does docs.rs need my help?

We have tried very hard to make sure that this change is backwards-compatible.
In other words, if you do nothing, the behavior will stay exactly the same as before:
documentation will be built for the default target (possibly taking into account `default-target` in the metadata),
and additionally will be built for all other tier one targets.
However, because of this, you need to opt-in in order for us to reduce our build costs.

We are considering turning this on by default in the future;
i.e. only building for one target unless multiple targets are specifically requested.
However, we do not want to break anyone's documentation, so in the meantime the feature is opt-in instead of opt-out.

## How can I learn more?

> - For more details about targets and what it means to be a tier-one target,

Reviewer: I'd prefer to avoid bullet points here.

By default, docs.rs builds all crates published to [crates.io] for [every tier one target][metadata].
However, most crates have the same content on all targets.
Of the platform-dependent crates, almost all target a single platform,
and do not need to be built on other targets.
For example, [`winapi`] only has documentation on the `x86_64-pc-windows-msvc`
and `i686-pc-windows-msvc` targets, and is blank on all others.

This feature allows you to request building only on specific targets.
For example, [`winapi`] could opt into only building windows targets
by putting the following in its `Cargo.toml`:

```toml
[package.metadata.docs.rs]
# This also sets the default target to `x86_64-pc-windows-msvc`
targets = ["x86_64-pc-windows-msvc", "i686-pc-windows-msvc"]
```

If you only need a single target, it's even simpler:

```toml
[package.metadata.docs.rs]
# This sets the default target to `x86_64-unknown-linux-gnu`
# and only builds that target
targets = ["x86_64-unknown-linux-gnu"]
```

See the [docs.rs documentation][metadata] for more details about how to opt-in.

## How does this help my crate?

Instead of building for every tier-one target, you can build for only a single target,

> reducing your documentation times by a factor of 6.

Reviewer: ```suggestion
reducing your documentation build times by a factor of 6.
```

that take several hours to document.

## How does this help docs.rs?

Building all crates from crates.io can take a long time!
Building fewer targets will allow us to reduce wait times for every crate.

> Additionally, this will decrease our fixed storage costs, allowing docs.rs to be sustainable into the future.

Reviewer: Reading this, it seems like docs.rs is not sustainable atm and this change is required for it to be sustainable.

```suggestion
Additionally, this will decrease the growth of our storage costs, improving the sustainability of the project.
```

## Possible future changes

We are considering turning this on by default in the future;
i.e. only building for one target unless multiple targets are specifically requested.

> However, we do not want to break anyone's documentation, so in the meantime the feature is opt-in instead of opt-out.

Reviewer: ```suggestion
However, we do not want to break anyone's documentation, so we're making this feature opt-in  while we decide the migration strategy.
```

This change will also make it easier for docs.rs to build
for targets that are not tier one, such as embedded targets.

## How can I learn more?

For more details about targets and what it means to be a tier-one target, see [platform support].
For details about the change, see the PR: [docs.rs#632].
For details about the motivation, see [docs.rs#343].

> For details about non-tier-one targets, see [docs.rs#563].

Reviewer: ```suggestion
You can learn more about the change in [the issue proposing it][docs.rs#343] and [the PR with the implementation][docs.rs#632]. Details on building non-tier-one targets are also available [in the issue requesting the feature][docs.rs#563].

More information on targets and what it means to be a tier-one target is available in the [platform support] page.
```
